### PR TITLE
require valid command

### PR DESCRIPTION
**Pull request by lrlna**
Check if `argv` provided is in `commandHandlers` if using `.strict()`.

A WIP pull for #174

### CODE REVIEW

**Comment by lrlna** [comment]
I'll open this up again tomorrow when I get the tests to pass :sweat_smile: :palm_tree: Will be better that way, I think :100:

**Comment by maxrimue** [comment]
The tests fail because the parsed command does not end up in the `argv` object, but in the `argv._` array, so `self.unknownArguments()` in `validation.js` does not check against it. Alternatively, you could create another check loop like this:
```javascript
argv._.forEach(function (key) {
  if (!commandKeys.hasOwnProperty(key)) {
    unknown.push(key)
  }
})
```
For me, it got the old tests and your new one passing, although I'm not sure if there's a better way to solve it, it's just what came to my mind right now.

File: lib/validation.js [hunk 1]
```diff
@@ -107,11 +108,18 @@ module.exports = function (yargs, usage, y18n) {
       if (key !== '$0' && key !== '_' &&
         !descriptions.hasOwnProperty(key) &&
         !demanded.hasOwnProperty(key) &&
+        !commandKeys.hasOwnProperty(key) &&
```
Comment (maxrimue) [inline lib/validation.js]: Great work, but isn't this line unnecessary now?

Comment (lrlna) [inline lib/validation.js]: I concur, it definitely is.

File: test/validation.js [hunk 2]
```diff
@@ -70,6 +70,23 @@ describe('validation tests', function () {
         .argv
     })
 
+    it('fails with invalid command', function (done) {
```
Comment (bcoe) [inline test/validation.js]: this is wonderful \o/

Comment (lrlna) [inline test/validation.js]: yay! :tada:

**Comment by lrlna** [comment]
@bcoe I think the next thing to implement within this scope would be the `.demand(1, ['name', 'age'])`. What do you think? Or should we merge this one step at a time?

File: lib/validation.js [hunk 3]
```diff
@@ -112,6 +113,12 @@ module.exports = function (yargs, usage, y18n) {
       }
     })
 
+    argv._.forEach(function (key) {
+      if (!commandKeys.hasOwnProperty(key)) {
+        unknown.push(key)
```
Comment (bcoe) [inline lib/validation.js]: I really appreciate that this implementation means that we don't need to go back and get new translations from people. Great work.

**Comment by bcoe** [comment]
@lrlna I agree, let's make it so that it's possible to specify a count for arguments, and a set of known arguments at the same time. `4.x` is coming along really well :)

**Comment by lrlna** [comment]
@bcoe coolest of beans; will get on that this weekend :ok_hand:

**Comment by lrlna** [comment]
Let me know what you folks think about this: I am thinking it may be a little more elegant doing something other than checking if max is an array. However, this way we are not committing any breaking changes (I don't think?).

File: test/usage.js [hunk 4]
```diff
@@ -58,6 +81,28 @@ describe('usage tests', function () {
           r.logs.should.have.length(0)
           r.exit.should.be.ok
         })
+        it('fails if one command and an argument not on the list are provided', function () {
```
Comment (bcoe) [inline test/usage.js]: could you add two additional test cases:

* one that verifies that it fails if all the required arguments exist, but not enough commands are provided?
* one that verifies that no failure occurs if the required arguments and the required number of commands are provided.

**Comment by bcoe** [comment]
This is 99% of the way there, added a note about two tests I'd love to see -- but otherwise this is looking wonderful.

File: test/usage.js (hunk 4 above)
Comment (lrlna) [inline test/usage.js]: will do! :100:

**Comment by lrlna** [comment]
@bcoe i got the usage tests :heavy_check_mark: ; still gotta got the validation tests in that we talked about.

**Comment by bcoe** [comment]
@lrlna looking great :+1:

**Comment by bcoe** [comment]
Once the Windows tests pass, I say :shipit:

**Pull request merged**